# List of tools

These are the tools that wrapped inside bacannot. **Cite** the tools whenever you use its output.

* [Refseq masher](https://github.com/phac-nml/refseq_masher) for bacterial species prediction
* [Prokka](https://github.com/tseemann/prokka) for generic annotation
* [barrnap](https://github.com/tseemann/barrnap) for rRNA prediction
* [mlst](https://github.com/tseemann/mlst) for classification within multi-locus sequence types (STs)
* [KofamScan](https://github.com/takaram/kofam_scan) for KO annotation
* [KEGGDecoder](https://github.com/bjtully/BioData/tree/master/KEGGDecoder) for drawing KO annotation
* [Nanopolish](https://github.com/jts/nanopolish) for methylation annotation
* [JBrowse](http://jbrowse.org/) for genome browser production
* [bedtools](https://bedtools.readthedocs.io/en/latest/) for annotation merging
* [gff-toolbox](https://github.com/fmalmeida/gff-toolbox) for plotting genomic islands
* [AMRFinderPlus](https://github.com/ncbi/amr/wiki), [ARGminer](https://bench.cs.vt.edu/argminer), [Resfinder](https://cge.cbs.dtu.dk/services/ResFinder/) and [RGI](https://github.com/arpcard/rgi) for antimicrobial genes annotation
* [PHASTER](https://phaster.ca/) database, [Phigaro](https://github.com/bobeobibo/phigaro) and [PhySpy](https://github.com/linsalrob/PhiSpy) for prophage sequences and genes prediction
* [IslandPath-DIMOB](https://github.com/brinkmanlab/islandpath) for genomic islands prediction
* [Plasmidfinder](https://cge.cbs.dtu.dk/services/PlasmidFinder/) and [Platon](https://github.com/oschwengers/platon) for _in silico_ plasmid detection
* [ICEberg](https://academic.oup.com/nar/article/47/D1/D660/5165266) database for the annotation of integrative and conjugative elements
* [digIS](https://github.com/janka2012/digIS) for detection of insertion sequences
* [Victors](http://www.phidias.us/victors/) and [VFDB](http://www.mgc.ac.cn/VFs/main.htm) for the annotation of virulence genes
